msal/src/Microsoft.Identity.Client/CacheV2/Impl/PathStorageWorker.cs

```diff
@@ -0,0 +1,551 @@
+                return StorageJsonUtils.CredentialFromJson(JObject.Parse(kvp.Value.Value.ToString()));
+            }
+
+            return null;
+        }
+
+        internal HashSet<string> SplitAndNormalizeScopes(string target)
+        {
+            var scopes = ScopeUtils.SplitScopes(target);
+            var normalizedScopes = new HashSet<string>();
+            foreach (string scope in scopes)
+            {
+                normalizedScopes.Add(NormalizeKey(scope));
+            }
+
+            return normalizedScopes;
+        }
+
+        private void ReadIdTokens(
+            string homeAccountId,
+            string environment,
+            string realm,
+            string clientId,
+            List<Credential> credentials)
+        {
+            string credentialPath = _credentialPathManager.GetCredentialPath(
+                homeAccountId,
+                environment,
+                realm,
+                clientId,
+                string.Empty,
+                CredentialType.OidcIdToken);
+            ReadCredential(credentialPath, credentials);
+        }
+
+        private void ReadCredential(string relativePath, List<Credential> credentials)
+        {
+            var json = Read(relativePath);
+            if (json != null && !json.IsEmpty())
+            {
+                credentials.Add(StorageJsonUtils.CredentialFromJson(json));
+            }
+        }
+
+        public JObject Read(string relativePath)
+        {
+            byte[] fileContent = _fileIo.Read(relativePath);
+            return fileContent.Length == 0 ? new JObject() : DecryptParse(fileContent, relativePath);
+        }
+
+        private JObject DecryptParse(byte[] fileContent, string relativePath)
+        {
+            if (fileContent == null || !fileContent.Any())
+            {
+                return new JObject();
+            }
+
+            try
+            {
+                byte[] decryptedFileContent = Decrypt(fileContent);
+                return JObject.Parse(Encoding.UTF8.GetString(decryptedFileContent, 0, decryptedFileContent.Length));
+            }
+            catch (JsonReaderException)
+            {
+            }
+            //catch (CryptoException)
+            //{
+
+            //}
+
+            return new JObject();
+        }
+
+        private string NormalizeKey(string data)
+        {
+            return data.ToLowerInvariant().Trim();
+        }
+
+        internal string GetCredentialPath(Credential credential)
+        {
+            return _credentialPathManager.GetCredentialPath(
+                credential.HomeAccountId,
+                credential.Environment,
+                credential.Realm,
+                credential.ClientId,
+                credential.FamilyId,
+                credential.CredentialType);
+        }
+
+        public void AddAccessTokenWithScopes(JObject accessTokens, JObject tokenToAdd)
+        {
+            if (!tokenToAdd.ContainsKey(StorageJsonKeys.Target))
```

Comment:
>StorageJsonKeys.Target [](start = 40, length = 22)

is this really the returned scopes? or what does target mean?

Reply:
The field in the json needs to be "target" and this property is named Target for consistency with what's being put there.  I think they named it Target instead of Scopes because the cache can store credentials that aren't OAuth2 compliant (and thus don't have scopes)